### PR TITLE
PARQUET-2157: add bloom filter fpp config

### PR DESCRIPTION
**Pull request by huaxingao**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-2157
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java [hunk 1]
```diff
@@ -39,13 +39,17 @@
 import java.io.File;
 import java.io.IOException;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.Callable;
 
 import net.openhft.hashing.LongHashFunction;
+import org.apache.commons.lang3.RandomStringUtils;
```
Comment (dongjoon-hyun) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:48]: Please add this as a test dependency to `parquet-hadoop/pom.xml`.
```
    <dependency>
      <groupId>org.apache.commons</groupId>
      <artifactId>commons-lang3</artifactId>
      <version>3.9</version>
      <scope>test</scope>
    </dependency>
```

**Comment by huaxingao** [comment]
cc @chenjunjiedada @ggershinsky @shangxinli

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java [hunk 2]
```diff
@@ -282,6 +286,63 @@ public void testParquetFileWithBloomFilter() throws IOException {
     }
   }
 
+  @Test
+  public void testParquetFileWithBloomFilterWithFpp() throws IOException {
+    final int totalCount = 100000;
```
Comment (chenjunjiedada) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:291]: Nit: Why do we need `final`?

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java [hunk 3]
```diff
@@ -282,6 +286,63 @@ public void testParquetFileWithBloomFilter() throws IOException {
...
+    for (int i = 0; i < testFpp.length; i++) {
+      File file = temp.newFile();
+      file.delete();
+      Path path = new Path(file.getAbsolutePath());
+      try (ParquetWriter<Group> writer = ExampleParquetWriter.builder(path)
+        .withPageRowCountLimit(10)
+        .withConf(conf)
+        .withDictionaryEncoding(false)
+        .withBloomFilterEnabled("name", true)
+        .withBloomFilterNDV("name", 100000l)
```
Comment (chenjunjiedada) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:316]: Nit: Can we use `TotalCount`?

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 2 above)
Comment (huaxingao) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java]: Removed.

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 3 above)
Comment (huaxingao) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java]: Fixed. Thanks!

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java [hunk 4]
```diff
@@ -282,6 +286,63 @@ public void testParquetFileWithBloomFilter() throws IOException {
...
+        int exist = 0;
+        while (distinctStrings.size() < totalCount) {
+          String str = RandomStringUtils.randomAlphabetic(10);
+          if (distinctStrings.add(str) &&
+            bloomFilter.findHash(LongHashFunction.xx(0).hashBytes(Binary.fromString(str).toByteBuffer()))) {
+            exist++;
+          }
+        }
+        // The exist should be less than totalCount * fpp. Add 10% here for error space.
+        assertTrue(exist < totalCount * (testFpp[i] * 1.1));
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:341]: just curious if `totalCount` is sufficient; how often `exist` > 0?

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java [hunk 5]
```diff
@@ -282,6 +286,63 @@ public void testParquetFileWithBloomFilter() throws IOException {
...
+
+      try (ParquetFileReader reader = ParquetFileReader.open(HadoopInputFile.fromPath(path, new Configuration()))) {
+        BlockMetaData blockMetaData = reader.getFooter().getBlocks().get(0);
+        BloomFilter bloomFilter = reader.getBloomFilterDataReader(blockMetaData)
+          .readBloomFilter(blockMetaData.getColumns().get(0));
+
+        // The exist counts the number of times FindHash returns true.
+        int exist = 0;
+        while (distinctStrings.size() < totalCount) {
+          String str = RandomStringUtils.randomAlphabetic(10);
```
Comment (ggershinsky) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:334]: the original values are 12 char long. Are we supposed to find a 10-char string among them?

File: parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java [hunk 6]
```diff
@@ -471,6 +484,12 @@ public Builder withBloomFilterNDV(String columnPath, long ndv) {
       return this;
     }
 
+    public Builder withBloomFilterFPP(String columnPath, double fpp) {
```
Comment (ggershinsky) [inline parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java:487]: what happens if this value is set, but the BF is not enabled? (general / per-column)

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 4 above)
Comment (ggershinsky) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:341]: Two related questions:
- what should be the `totalCount` to reliably ensure that a) `exist > 0` b) `exist < totalCount * (testFpp[i] * 1.1)` ? Depending on the `fpp` value, we can get a random assert exception if `totalCount` is too low (also, `exist` could be just 0 then). If `totalCount` is high, the unitest could take a very long time. 
- how long does this unitest run on your laptop? (with the current `totalCount` of 100000).

Comment (huaxingao) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:342]: Basically `exist` > 0 is false positive. which happens when any given hash value that was never inserted into the bloom filter causes the check to return true. I don't think there is a simple closed-form calculation of this probability, but setting `totalCount` to be `100000` seems to be a pretty safe number for the test to pass.

I am thinking we probably should disallow the Bloom filter's size to be unreasonably small. We currently only have the 
maximum bytes of the Bloom filter. Shall we also have the minimum bytes of the Bloom filter? What do you think? @chenjunjiedada 

The test takes about 2300 milli seconds on my laptop.

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 5 above)
Comment (huaxingao) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java]: Changed. Thanks!

File: parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java (hunk 6 above)
Comment (huaxingao) [reply parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java:487]: This value will be silently ignored.

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 4 above)
Comment (chenjunjiedada) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java:342]: The size of the bloom filter is computed with `ndv` and `fpp`. So even the size is "unreasonable" small it should be enough to handle the given situation. Right?

**Comment by ggershinsky** [comment]
> The test takes about 2300 milli seconds on my laptop.

Ok, this is reasonable. If this time is sufficient for reliably testing the upper limit of FPPs, it should be good enough to also check the lower limit, eg `exist > totalCount * (testFpp[i] * 0.9)` , or `exist > totalCount * (testFpp[i] * 0.5)` , or even `exist > 0`. What do you think? This way, we'll be certain the test passes not because `exist` is just 0.

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java (hunk 4 above)
Comment (huaxingao) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetWriter.java]: Yes. Agree.

**Comment by huaxingao** [comment]
> it should be good enough to also check the lower limit, eg exist > totalCount * (testFpp[i] * 0.9) , or exist > totalCount * (testFpp[i] * 0.5) , or even exist > 0. What do you think? This way, we'll be certain the test passes not because exist is just 0.

Thanks for the suggestion! I can't find a reliable number for the lower limit. I put `exist > 0`.

**Review by ggershinsky — APPROVED** [review]

**Comment by shangxinli** [comment]
LGTM

**Comment by huaxingao** [comment]
Thank you all very much! @chenjunjiedada @dongjoon-hyun @ggershinsky @shangxinli

**Comment by dongjoon-hyun** [comment]
Could you resolve JIRA please? I realized that the JIRA is still open although this is delivered.
- https://issues.apache.org/jira/browse/PARQUET-2157

**Pull request merged**